### PR TITLE
Auth Next changes in SDK

### PR DESCRIPTION
**Pull request by dmkozh**
### What

Adapt Soroban SDK to the host changes introduced in https://github.com/stellar/rs-soroban-env/pull/645

Overview of the changes:

- Replace the `AccountId`, `invoker`, and `soroban-auth` library with the single `Address` type that supports host-based authentication and authorization.
- Use the recording auth for testing and provide a simple utility to verify that authorization has happened. Signatures are no longer required for tests thanks to that.
- Replace network passphrase with id (this just reflects the host-side change; we don't really need to have both available to contrats)
- Introduced a stub for the account contract SDK

### Why

Auth Next motivation is described in the env PR. From the SDK perspective the auth and testing are greatly simplified.

### Known limitations

Some of the possible followups:

- The account SDK would need to be developed more to provide a better support for building the custom account contracts
- Strkey should be used for displaying the addresses
- Add testing utility to verify non-top-level authorizations 
- Add some macros to generate the auth code for simple cases (even though this is just a line of code, it would be more readable to have an explicit function annotation that requires a given address to authorize a call with all the input arguments).
- Maybe bring back some of the classic account related functionality for the integration testing. While this should be unnecessary for the unit tests, there may be some need for it in snapshot-based testing. Not quite certain about that yet though.

### CODE REVIEW

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:223]: It's unfortunate that we have to have both of these functions. As a user I'm not sure which one I should use. I'm likely to choose the ID in situations where I should use the Adddress. I'm not sure what the solution to that is though.

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:383]: It seems a bit strange to me that we'd modify the env in this way, rather than use the pattern we have used for storage where we provide an implementation, and there is a recording or enforcing implementation. Is there a reason we can't use that existing pattern here?

File: soroban-sdk/src/accounts.rs
Comment (leighmcculloch) [inline soroban-sdk/src/accounts.rs:402]: How do folks write tests with accounts without this? How do I create an account, with a particular balance, and use that balance in a test? Is there a new way to do this?

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:558]: If we kept the Accounts testutils we could use them to create this account rather than embedding it into the registration.

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:520]: It seems inconsistent that `register_stellar_asset_contract` takes an `asset`, and `register_stellar_asset_contract_with_admin` randomly generates an asset. I think inconsistencies like this can make APIs confusing to use so we should address that for usability.

Could we allow the asset to be passed in, and add a utility for randomly generating an asset if that functionality is required?

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:629]: The side effects of calling a "verify" function are surprising to me, and I think will be surprising to other people. Functions that verify things are usually idempotent. See below comment for a suggestion.

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:663]: This type of testing reminds me of tests that utilize mocks, where we're trying to track if a call to the mock has occurred or not. Here we're trying to track if a call to require_authorization has occurred.

I think this testing API would be more intuitive to users if we presented it as test double or a mock or a spy, since that's effectively what it is.

But, I think we should try to find another way to make this assertion, because the error this assertion displays will be difficult to debug because it is just a yes or a no.

Could we add a new function that gets a Vec<> of require authorizations that have been called? The way this would be written in a test would then be something like this:

```rust
client.transfer(&address, &1000_i128);
assert_eq!(
    env.auth().all(),
    std::vec![
        (&address, &contract_id, "transfer", (500_i128,).into_val(&env)),
    ],
);
```

Writing the test this way has several advantages:
- If someone doesn't know what is going on they can print out all the authorizations.
- The assert_eq will print out the full list of authorizations if they don't match what is expected.

As a bonus the API of the SDK will now only present facts and data, and is less prescriptive, and causes less side-effects.

**Review by leighmcculloch — COMMENTED** [review]
This is looking really nice. I have a few asks inline.

File: soroban-sdk/src/env.rs
Comment (dmkozh) [reply soroban-sdk/src/env.rs:558]: Agree, but I'm not sure if we should keep these as is, as there is really no other use for them yet. Let's clean this up when we figured out when/if the accounts may be needed.

File: soroban-sdk/src/env.rs
Comment (dmkozh) [reply soroban-sdk/src/env.rs:520]: I don't really want to expose the assets to the users at all; this is meant as a simple way to register a built-in token for your tests without worrying about any of the classic semantics. We could maybe even name this in a less specific way (`register_test_token`?)

I'm actually not convinced `register_stellar_asset_contract` is useful as a public now; maybe we could hide it for now and see if someone really needs it?

File: soroban-sdk/src/env.rs
Comment (dmkozh) [reply soroban-sdk/src/env.rs:663]: I have mixed feelings about setting up mocks; while pretty nice in concept they're frequently tricky to setup correctly and require pretty significant library support.

I like your suggestion. The current design stems from the previous iteration where you needed to pass the whole call stack and I want to allow users to only verify a few calls they could be interested in. But If we do this for the top level, that's probably simple enough to use.

Since this requires a host change, I'll open an issue for that for now: https://github.com/stellar/rs-soroban-env/issues/663

File: soroban-sdk/src/env.rs
Comment (dmkozh) [reply soroban-sdk/src/env.rs:629]: That's a good suggestion, thanks. Replied there.

File: soroban-ledger-snapshot/src/lib.rs
Comment (leighmcculloch) [reply soroban-ledger-snapshot/src/lib.rs:30]: Do you mean the `set_ledger_info` function in env? We should be able to sha256 hash the value when passing in, or let the input to that function be the passphrase and hash inside the function?

File: soroban-sdk/src/address.rs
Comment (leighmcculloch) [reply soroban-sdk/src/address.rs]: The strkey crate supports contracts already, it's the `C` strkey. See https://docs.rs/stellar-strkey/latest/stellar_strkey/struct.Contract.html.

File: soroban-sdk/src/env.rs
Comment (leighmcculloch) [reply soroban-sdk/src/env.rs:520]: In that case I think what we should do for this release is remove the function that accepts an asset, and make this function the only function for creating an asset contract. The name can be shortened, the with_admin is not really necessary when the function is on its own.

```suggestion
    pub fn register_stellar_asset_contract(&self, admin: Address) -> BytesN<32> {
```

I'd stick with the stellar asset contract name over test token, just for building that ubiquitous language around what it is vs what a token is.

Wdyt?

File: soroban-ledger-snapshot/src/lib.rs
Comment (dmkozh) [reply soroban-ledger-snapshot/src/lib.rs:30]: I mean `set_ledger_info` in `LedgerSnapshot`. It uses `LedgerInfo` to set the fields of the snapshot, but the `LedgerInfo` itself is defined in host and doesn't have passphrase, so we can't recover it.

If you think passphrase significantly improves the UX we can bring it back in host as well, though I'm still not sure about exposing both as the host fns.

File: soroban-sdk/src/address.rs
Comment (dmkozh) [reply soroban-sdk/src/address.rs]: Nice, we've apparently just used the previous version, which is why I haven't noticed it.

File: soroban-sdk/src/address.rs
Comment (leighmcculloch) [inline soroban-sdk/src/address.rs:56]: It's supported: http://docs.rs/stellar-strkey/latest/stellar_strkey/struct.Contract.html

File: soroban-sdk/src/accounts.rs
Comment (dmkozh) [reply soroban-sdk/src/accounts.rs:402]: The real question here is what would be the reason for creating an account with balance. So far I don't see one - you can just create a token with a given admin and use it to mint the balance (actually I think we'll be able to get rid even of the admin; thanks to the recording auth we can just mint in tests without providing any credentials, see https://github.com/stellar/rs-soroban-env/issues/653 for details). I think this is a much more natural UX, given that it doesn't go beyond what contracts already do.
As I've already mentioned, for the snapshot-based tests I would *hope* that the accounts/trustlines would be set up as a part of the snapshot itself. But if not, we can bring this back (maybe with a bit different API).

File: soroban-sdk/src/testutils.rs
Comment (leighmcculloch) [inline soroban-sdk/src/testutils.rs:168]: I think I was misunderstood. For consistency with the other types the impl should live with the type, and there should be a trait that lives here with the same name that is being implemented. Have a look at the traits above and their respective implementations for how they work with their type.

File: soroban-sdk/src/address.rs
Comment (leighmcculloch) [inline soroban-sdk/src/address.rs:178]: Usually the sections like Panic follow after the discussion paragraph.
```suggestion
    /// During the on-chain execution the Soroban host will perform the needed
    /// authentication (verify the signatures) and ensure the replay prevention.
    /// The contracts don't need to perform this tasks.
    ///
    /// The arguments don't have to match the arguments of the contract
    /// invocation. However, it's considered the best practice to have a
    /// well-defined, deterministic and ledger-state-independent mapping between
    /// the contract invocation arguments and `require_auth` arguments. This
    /// will allow the contract callers to easily build the required signature
    /// payloads and prevent potential authorization failures.
    ///
    /// When called in the tests, the `require_auth` calls are just recorded and
    /// no signatures are required. In order to make sure that the contract
    /// has indeed called `require_auth` for this Address with expected arguments
    /// use `env.verify_top_authorization`.
    ///
    /// ### Panics
    ///
    /// If the invocation is not authorized.
```

**Review by leighmcculloch — COMMENTED** [review]
I think my suggestions were misunderstood in a couple places.

I think the main things left are:

1. Figuring out if we can keep network passphrase, or if we need to move to network id in the SDK. I agree network id should be used within the env.

2. Can we replace the verify_top_auth function.

It would be good to figure out (1) before merging so that we don't flip-flop the network passphrase code. But it's also fine to flip flop it if we can't figure it out.

I think we can fix (2) post-merge, but I think it is important we fix that before release, so maybe worth doing in this PR too.

File: soroban-sdk/src/env.rs
Comment (dmkozh) [reply soroban-sdk/src/env.rs]: Sounds good, renamed. Also updated `use_token_contract` tests - now that I think about it, there is really no good reason for them to try covering different asset types; that's the host logic and its already covered in host.

File: soroban-sdk/src/address.rs
Comment (dmkozh) [reply soroban-sdk/src/address.rs]: Updated.

File: soroban-ledger-snapshot/src/lib.rs
Comment (leighmcculloch) [reply soroban-ledger-snapshot/src/lib.rs:30]: Oh that's a good point. Leave it for now, I don't think the UX improvement warrants the inconvenience of cleaning this up. I'll create an issue for remapping passphrase in the ledger snapshot and SDK, but I think it's lower priority.

Comment (leighmcculloch) [reply soroban-ledger-snapshot/src/lib.rs:30]: https://github.com/stellar/rs-soroban-sdk/issues/851

File: soroban-sdk/src/accounts.rs
Comment (leighmcculloch) [reply soroban-sdk/src/accounts.rs:402]: Ok, lets see how far we can get without it, but we've been here before where we left it out, then later added it, etc. If we don't need it though, it will make for less code to maintain.

File: soroban-sdk/src/testutils.rs
Comment (dmkozh) [reply soroban-sdk/src/testutils.rs]: Oh, sorry, I can't read. Updated for consistency.

FWIW this reminded me that the current approach is pretty IDE-unfriendly. Not only you won't get these functions in auto-completion by default, but also rust analyzer fails to properly import it (it doesn't insert `as _` automatically). So we actually may think of revisiting this (not in this PR obviously).

**Comment by dmkozh** [comment]
> I think we can fix (2) post-merge, but I think it is important we fix that before release, so maybe worth doing in this PR too.

I would suggest merging this as is (given everything else looks ok) in order to unblock your changes or any other unrelated changes. I'll prioritize this to make sure it makes it into release.

File: soroban-sdk/src/accounts.rs
Comment (dmkozh) [reply soroban-sdk/src/accounts.rs:402]: I want to believe that the testing without accounts is much more plausible now (I think getting rid of invoker goes a long way in enabling this). But let's see indeed.

File: soroban-sdk/src/testutils.rs
Comment (leighmcculloch) [reply soroban-sdk/src/testutils.rs]: > rust analyzer fails to properly import it (it doesn't insert `as _` automatically

+1 We could rename the traits to AddressTestutils.

File: soroban-sdk/src/address.rs
Comment (leighmcculloch) [inline soroban-sdk/src/address.rs:213]: The cfg's should be as any test and testutils I think, although it might not matter. That's what we used in other impl sections like this:

```suggestion
#[cfg(any(test, feature = "testutils"))]
use crate::env::xdr::{Hash, ScAddress, ScObject};
#[cfg(any(test, feature = "testutils"))]
use crate::{testutils::random, BytesN};
#[cfg(any(test, feature = "testutils"))]
```

Comment (leighmcculloch) [inline soroban-sdk/src/address.rs:213]: The cfg's should be as any test and testutils I think, although it might not matter. That's what we used in other impl sections like this:

```suggestion
#[cfg(any(test, feature = "testutils"))]
use crate::env::xdr::{Hash, ScAddress, ScObject};
#[cfg(any(test, feature = "testutils"))]
use crate::{testutils::random, BytesN};
#[cfg(any(test, feature = "testutils"))]
```

File: soroban-sdk/src/testutils.rs
Comment (leighmcculloch) [inline soroban-sdk/src/testutils.rs:161]: Why do we need the wasm cfg?

**Review by leighmcculloch — APPROVED** [review]
Looks good, +1 to following up with the remaining things.

Two things with cfgs stand out to me which I commented on inline, just because they look inconsistent with other code following the same pattern. I'll try changing them and if they work changed I'll merge. Otherwise I'll merge them as it is.

File: soroban-sdk/src/testutils.rs
Comment (leighmcculloch) [reply soroban-sdk/src/testutils.rs:161]:
```suggestion
```

**Comment by leighmcculloch** [comment]
@dmkozh I pushed a commit before merging. See https://github.com/stellar/rs-soroban-sdk/pull/848/commits/1ae8ac0063ed0309c88ade6a726b123020c4f078.

File: soroban-sdk/src/testutils.rs
Comment (dmkozh) [reply soroban-sdk/src/testutils.rs]: I added this because `ScVal` conversions are conditioned on `not(wasm)`, while both of the functions utilize these. So IIUC the build won't compile in test/testutils mode with wasm target. That's probably not a big deal, given that we don't normally build this configuration, but I think it would be nice to have all the feature combinations and archs to compile. Maybe `ScVal` conversions should be gated by the test/testutils and not target?

Comment (leighmcculloch) [reply soroban-sdk/src/testutils.rs]: Yeah it seems like overtime we've ended up with two ways to specify the same thing. This came about because originally the sdk was going you be used for native contracts too. In native contracts ScVal conversions would be supported, but not testutils.

Note that it's not possible to enable testutils on Wasm builds due you the condition below. So while none of this is perfect, functionally there's no impact.

https://github.com/stellar/rs-soroban-sdk/blob/e7910c7057088b3418d84af04f2c04313d0c756f/soroban-sdk/src/lib.rs#L45-L46

Comment (dmkozh) [reply soroban-sdk/src/testutils.rs]: Cool, then this is good enough for now.

**Pull request merged**
